### ISSUE TITLE
S3  putBucketLifecycleConfiguration OperationAborted error

### ISSUE THREAD

**Issue by Dayjo** [aws/aws-sdk-js#1091]
Hello!

TL;DR

The reason I was getting this, was because it seems using both the callback function AND the promise causes it to get run more than once, i.e.

```js
var prom = S3.putBucketLifecycleConfiguration(params, function(err, data) {
     // error handling
}).promise();
```

Should just be; 
```
var prom = S3.putBucketLifecycleConfiguration(params).promise();
prom.then(function(){
    // Success
}, function(err){ 
    // error handling
});
```

------
------

I have a Lambda running on Node which creates an S3 bucket and then proceeds to put a lifecycle configuration, then a notification configuration onto the bucket.

I'm getting the following error which is causing the lambda to stop, and the next part of the code (notification configuration) to not run. Http response below; 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Error>
<Code>OperationAborted</Code>
<Message>A conflicting conditional operation is currently in progress against this resource. Please try again.</Message>
<RequestId>D3E608F2A34D435F</RequestId>
<HostId>jjovfq3lwfMBH2jQxswxvkNQKSr/HfwG+knxSxiqXgrP30fTu9zMHJrtyOk7yJXVrorziHJuzEo=</HostId>
</Error>
```

from the following function;

```js
var params = {
            Bucket: bucketName,
            LifecycleConfiguration: {
                Rules: [
                    // DELETE old uploads that have failed to be processed
                    {
                        ID: 'deleteOldUploads',
                        Prefix: 'uploads/',
                        Status: 'Enabled',
                        AbortIncompleteMultipartUpload: {
                            DaysAfterInitiation: 1
                        },
                        Expiration: {
                            Days: 2,
                        }
                    },

                    // Move non current files into infrequent access
                    {
                        ID: 'infrequentAccess',
                        Prefix: '_store/', /* required */
                        Status: 'Enabled', /* required */
                        NoncurrentVersionTransitions: [
                            {
                                NoncurrentDays: 30,
                                StorageClass: "STANDARD_IA"
                            }
                        ],
                    },
                    /* more items */
                ]
            }
        };

var lifecycle = S3.putBucketLifecycleConfiguration(params, function(err, data) {
            if (err) {
                console.log("LIFECYCLE ERROR");
                console.log(this.httpResponse.body.toString());
                console.log(err, err.stack); // an error occurred
            }
}).promise();

lifecycle.then(function{
     // Do the Notification Configuration..
```

One thing to **note** is that the lifecycle configurations actually appear fine in the bucket, but because the lambda stops, the notification config doesn't, and also other code doesn't execute. The error tends to occur a few times. 

---- 

The bucket definitely exists at the point it runs as I'm using the `waitFor` function and outputting the result of that before running the putBucketLifecycleConfiguration function. Also, the bucket exists, WITH the lifecycle config despite the error, just stops the other code from running. If I pull out this function, the Notification Config carries on as normal and everything is fine. 

I am also using Promises within a function to chain the commands as I thought perhaps they were trying to run at the same time but shouldn't be.

The strange thing is that I'm getting the error multiple times and the second time there's no `response.body` as per my logging above as per this cloudwatch screenshot;
![](http://c.dayjo.me/0A1C2U1W2U1k/Image%202016-08-10%20at%205.55.03%20pm.png)

I've been unable to find out all of the causes for this error, most of the places I've looked up (stackoverflow, various issues on here) they seem to be related to creating a bucket rather than putting a lifecycle config.

If there's any more info I can provide just let me know!

**Comment by chrisradek** [comment]
@Dayjo 
The way you're making use of promises in the SDK is a little off.

> ```javascript
> var lifecycle = S3.putBucketLifecycleConfiguration(params, function(err, data) {
>             if (err) {
>                 console.log("LIFECYCLE ERROR");
>                 console.log(this.httpResponse.body.toString());
>                 console.log(err, err.stack); // an error occurred
>             }
> }).promise();
> ```

When you're using `.promise()`, you don't want to specify a callback function. Doing so may actually cause your operation to be called twice under some conditions!

Instead, you should revise your code to handle errors differently:
```javascript
var lifecycle = S3.putBucketLifecycleConfiguration(params).promise();
lifecycle.then(function(data) {
  /* Do the notification config */
}).catch(function(err) {
  console.log('LIFECYCLE ERROR');
  console.log(err);
});
```

That `catch` will capture any errors in your promise chain. You could also provide a second callback function to `then()` that would handle an error, though the former approach is generally better since it handles all errors.

Can you try making these changes and let us know if you're still seeing this issue?

**Comment by Dayjo** [comment]
Hah :) got there JUST after I figured that out, hah have updated the top post, but yes, this is because of the incorrect use of promises. I didn't realise/think/take on board that you could only use one :)
